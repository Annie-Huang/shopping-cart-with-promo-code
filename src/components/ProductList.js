import React,  { useEffect } from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import * as productActions from "../redux/actions/productActions";
import * as discountCodeActions from "../redux/actions/discountCodeActions";
import * as shoppingCartActions from "../redux/actions/shoppingCartActions";
import Spinner from "./common/Spinner";
import styles from "./ProductList.module.css";
import toastr from 'toastr/toastr';

export const ProductList = ({
    products,
    loadProducts,
    updateItemInCart,
    discountCodes,
    loadDiscountCodes,
}) => {
    useEffect(() => {
        loadProducts();
        // loadProducts().catch(error => {
        //     alert("Loading products failed" + error);
        // });
    }, [loadProducts]);

    useEffect(() => {
        loadDiscountCodes();
        // loadDiscountCodes().catch(error => {
        //     alert("Loading discount codes failed" + error);
        // });
    }, [loadDiscountCodes]);

    const updateShoppingCart = (data) => {
        updateItemInCart(data);
        const action = data.quantity === 1 ? 'Add' : 'Remove';
        toastr.success(`${action} one ${data.product.name} success`);
    };

    const productList = products.map(product =>
        <div className="card" key={product.id}>
            <div className="card-header text-center text-white bg-primary">
                <h4 className="card-title">{product.name}</h4>
            </div>
            <div className="card-body">
                <div className='row'>
                    <div className='col-md-3'>Price:</div>
                    <div className='col-md-7'>${product.price}</div>
                </div>
            </div>
            <div className="card-footer">
                <button type="button"
                        className="btn btn-primary"
                        onClick={() => updateShoppingCart({product, quantity: 1})}
                >
                    Add 1 item
                </button>
                &nbsp;&nbsp;
                {product.productInCart &&
                <button type="button"
                        className="btn btn-secondary"
                        onClick={() => updateShoppingCart({product, quantity: -1})}
                >
                    Remove 1 item
                </button>
                }
            </div>
        </div>
    );

    const discountCodeList = discountCodes.map(discountCode =>
        <div key={discountCode.id}>{discountCode.id} : {discountCode.description}</div>
    );

    return (
        products.length === 0 || discountCodes.length === 0 ?
            <Spinner/> :
            <>
                <div className="card-deck">
                    {productList}
                </div>
                <br/>
                <div className={styles.warning}>Test CSS Modules</div>
                <h5>Available discount codes are as following. Please keep in mind that you can only apply for one.</h5>
                {discountCodeList}
            </>
    )
};

ProductList.propTypes = {
    products: PropTypes.array.isRequired,
    loadProducts: PropTypes.func.isRequired,
    updateItemInCart: PropTypes.func.isRequired,
    discountCodes: PropTypes.array.isRequired,
    loadDiscountCodes: PropTypes.func.isRequired,
};

export const updateProductsWithInCartInfo = (products, cartItems) => {
    const updatedProducts = [];
    products.forEach(product => {
        const matchCartItem = cartItems.find((cartItem) => cartItem.product.id === product.id);
        updatedProducts.push({...product, productInCart: !!matchCartItem});
    });

    return updatedProducts;
};

const mapStateToProps = (state, ownProps) => ({
    products: updateProductsWithInCartInfo(state.products, state.cartItems),
    discountCodes: state.discountCodes,
});

const mapDispatchToProps = {
    loadProducts: productActions.loadProducts,
    updateItemInCart: shoppingCartActions.updateItemInCart,
    loadDiscountCodes: discountCodeActions.loadDiscountCodes,
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ProductList);
