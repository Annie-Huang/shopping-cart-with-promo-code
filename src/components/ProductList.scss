.scss-warning {
  background-color: #ff9700;
}
