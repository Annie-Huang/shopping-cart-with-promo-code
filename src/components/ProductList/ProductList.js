import React,  { useEffect } from 'react';
import {connect} from 'react-redux';
import * as productActions from "../../redux/actions/productActions";
import PropTypes from 'prop-types';

export const ProductList = ({
    products,
    loadProducts
}) => {
    useEffect(() => {
        loadProducts().catch(error => {
            alert("Loading courses failed" + error);
        });
    }, []);

    return (
        <div>
            products.size: {products.length}
        </div>
    );
};

ProductList.propTypes = {
    products: PropTypes.array.isRequired,
    loadProducts: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
    products: state.products,
});

const mapDispatchToProps = {
    loadProducts: productActions.loadProducts
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ProductList);
