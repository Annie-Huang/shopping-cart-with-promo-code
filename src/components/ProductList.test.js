import React from 'react';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import {ProductList, updateProductsWithInCartInfo} from './ProductList';

describe('ProductList presentation', () => {
    const loadProducts = jest.fn();
    const updateItemInCart = jest.fn();
    const loadDiscountCodes = jest.fn();
    const products = require('../resources/fixtures/products.json');
    const discountCodes = require('../resources/fixtures/discountCodes.json');
    products[0].productInCart = true;
    products[1].productInCart = false;
    products[2].productInCart = true;
    const props = {
        products,
        loadProducts,
        updateItemInCart,
        discountCodes,
        loadDiscountCodes
    };

    it('should render ProductList component', () => {
        expect(toJson(shallow(<ProductList {...props} />))).toMatchSnapshot();
    });

    // Need to find a way to directly call method for stateless component in react....
    xit('#updateItemInCart -- called directly', () => {
        const wrapper = shallow(<ProductList {...props} />).instance();
        const data = {
            product: products[0],
            quantity: 1
        };
        console.log('wrapper.updateShoppingCart=', wrapper.updateShoppingCart({}));
        wrapper.updateShoppingCart(data);

        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateItemInCart should be called when clicking Add 1 item button', () => {
        const wrapper = shallow(<ProductList {...props} />);

        const buttons = wrapper.find('button');
        const addButton =  buttons.first();
        expect(addButton.text()).toBe('Add 1 item');

        addButton.simulate('click');
        const data = {
            product: products[0],
            quantity: 1
        };
        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateItemInCart should be called when clicking Remove 1 item button', () => {
        const wrapper = shallow(<ProductList {...props} />);

        const buttons = wrapper.find('button');
        const removeButton =  buttons.at(1);
        expect(removeButton.text()).toBe('Remove 1 item');
        removeButton.simulate('click');

        const data = {
            product: products[0],
            quantity: -1
        };
        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateProductsWithInCartInfo should be productInCart property into each product', () => {
        const originalProducts = require('../resources/fixtures/products.json');
        const cartItems = require('../resources/fixtures/cart-items.json');

        const updatedProducts = updateProductsWithInCartInfo(originalProducts, cartItems);
        expect(updatedProducts[0].productInCart).toBe(true);
        expect(updatedProducts[1].productInCart).toBe(true);
        expect(updatedProducts[2].productInCart).toBe(false);
    });

});
