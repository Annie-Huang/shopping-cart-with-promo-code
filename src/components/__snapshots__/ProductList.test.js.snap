// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList presentation should render ProductList component 1`] = `
<Fragment>
  <div
    className="card-deck"
  >
    <div
      className="card"
      key="wf"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Workflow
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            199.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Remove 1 item
        </button>
      </div>
    </div>
    <div
      className="card"
      key="docgen"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Document Generation
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            9.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
      </div>
    </div>
    <div
      className="card"
      key="form"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Form
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            99.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Remove 1 item
        </button>
      </div>
    </div>
  </div>
  <br />
  <div
    className="warning"
  >
    Test CSS Modules
  </div>
  <h5>
    Available discount codes are as following. Please keep in mind that you can only apply for one.
  </h5>
  <div
    key="RRD4D32"
  >
    RRD4D32
     : 
    10% discount for orders above $1000 (pre-discount)
  </div>
  <div
    key="44F4T11"
  >
    44F4T11
     : 
    15% discount for orders above $1500 (pre-discount)
  </div>
  <div
    key="FF9543D1"
  >
    FF9543D1
     : 
    Reduces the docgen price to $8.99 a unit when at least 10 documents are purchased
  </div>
  <div
    key="YYGWKJD"
  >
    YYGWKJD
     : 
    Reduces the form price to $89.99 a unit when at least 1 wf is purchased
  </div>
</Fragment>
`;
